src/tpm2/crypto/openssl/CryptSym.c

```diff
@@ -458,6 +461,262 @@ CryptSymmetricDecrypt(
+		      TPM2B_IV            *ivInOut,       // IN/OUT: IV for decryption.
+		      TPM_ALG_ID           mode,          // IN: Mode to use
+		      INT32                dSize,         // IN: data size (may need to be a
+		      //     multiple of the blockSize)
+		      const BYTE          *dIn            // IN: data buffer
+		      )
+{
+    INT16                blockSize;
+    BYTE                *iv;
+    BYTE                 defaultIv[MAX_SYM_BLOCK_SIZE] = {0};
+    evpfunc              evpfn;
+    EVP_CIPHER_CTX      *ctx = NULL;
+    int                  outlen1, outlen2;
+    BYTE                *pOut = dOut;
+    BYTE                *buffer = NULL;
+    UINT32               buffersize = 0;
+    BYTE                 keyToUse[MAX_SYM_KEY_BYTES];
+    UINT16               keyToUseLen = (UINT16)sizeof(keyToUse);
+    TPM_RC               retVal = TPM_RC_SUCCESS;
+
+    // These are used but the compiler can't tell because they are initialized
+    // in case statements and it can't tell if they are always initialized
+    // when needed, so... Comment these out if the compiler can tell or doesn't
+    // care that these are initialized before use.
+    pAssert(dOut != NULL && key != NULL && dIn != NULL);
+    if(dSize == 0)
+	return TPM_RC_SUCCESS;
+    TEST(algorithm);
+    blockSize = CryptGetSymmetricBlockSize(algorithm, keySizeInBits);
+    if(blockSize == 0)
+	return TPM_RC_FAILURE;
+    // If the iv is provided, then it is expected to be block sized. In some cases,
+    // the caller is providing an array of 0's that is equal to [MAX_SYM_BLOCK_SIZE]
+    // with no knowledge of the actual block size. This function will set it.
+    if((ivInOut != NULL) && (mode != ALG_ECB_VALUE))
+	{
+	    ivInOut->t.size = blockSize;
+	    iv = ivInOut->t.buffer;
+	}
+    else
+	iv = defaultIv;
+
+    evpfn = GetEVPCipher(algorithm, keySizeInBits, mode, key,
+                         keyToUse, &keyToUseLen);
+    if (evpfn ==  NULL)
+        return TPM_RC_FAILURE;
+
+    if (dIn == dOut) {
+        // in-place encryption; we use a temp buffer
+        buffersize = TPM2_ROUNDUP(dSize, blockSize);
+        buffer = malloc(buffersize);
+        if (buffer == NULL)
+            ERROR_RETURN(TPM_RC_FAILURE);
+        pOut = buffer;
+    }
+
+#if ALG_TDES && ALG_CTR
+    if (algorithm == TPM_ALG_TDES && mode == ALG_CTR_VALUE) {
+        TDES_CTR(keyToUse, keyToUseLen * 8, dSize, dIn, iv, pOut, blockSize);
+        if (pOut != dOut)
+            memcpy(dOut, pOut, dSize);
+        ERROR_RETURN(TPM_RC_SUCCESS);
+    }
+#endif
+
+    ctx = EVP_CIPHER_CTX_new();
+    if (!ctx)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    if (EVP_DecryptInit_ex(ctx, evpfn(), NULL, keyToUse, iv) != 1)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    EVP_CIPHER_CTX_set_flags(ctx, EVP_CIPH_NO_PADDING);
+    if (EVP_DecryptUpdate(ctx, pOut, &outlen1, dIn, dSize) != 1)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    EVP_CIPHER_CTX_set_flags(ctx, EVP_CIPH_NO_PADDING);
```

Comment:
repeated as in encryption

Reply:
... and not necessary in the decryption part.